core/src/main/java/org/apache/carbondata/core/cache/dictionary/DoubleArrayTrieDictionary.java

```diff
@@ -0,0 +1,384 @@
+    int from = start;
+    int newKey = bKey;
+    TreeSet<Integer> children = getChildren(from);
+    children.add(new Integer(newKey));
+    int newBasePos = findFreeRoom(children);
+    children.remove(new Integer(newKey));
+
+    int oldBasePos = base[start];
+    base[start] = newBasePos;
+
+    int oldPos, newPos;
+    for (Integer child : children) {
+      oldPos = oldBasePos + child;
+      newPos = newBasePos + child;
+      if (oldPos == from) from = newPos;
+      base[newPos] = base[oldPos];
+      check[newPos] = check[oldPos];
+      if (newPos >= size) size = newPos + 1;
+      if (base[oldPos] > 0) {
+        TreeSet<Integer> cs = getChildren(oldPos);
+        for (Integer c : cs) {
+          check[base[oldPos] + c] = newPos;
+        }
+      }
+      base[oldPos] = EPTY_BACK_VALUE;
+      check[oldPos] = EPTY_BACK_VALUE;
+    }
+    return from;
+  }
+
+  /**
+   * Insert element (byte[]) into DAT.
+   * 1. if the element has been DAT then return.
+   * 2. if position which is empty then insert directly.
+   * 3. if conflict then resolve it.
+   *
+   * @param bKeys
+   * @return
+   */
+  private boolean insert(byte[] bKeys) {
+    int from = 1;
+    int klen = bKeys.length;
+    for (int i = 0; i < klen; i++) {
+      int c = bKeys[i] & 0xFF;
+      int to = base[from] + c;
+      reSize((int) (to * 1.2) + 1);
+      if (check[to] == from) {
+        if (i == klen - 1) return true;
+        from = to;
+      } else if (check[to] == EPTY_BACK_VALUE) {
+        check[to] = from;
+        if (i == klen - 1) {
+          base[to] = -id;
+          id = id + 1;
+          return true;
+        } else {
+          int next = bKeys[i + 1] & 0xFF;
+          base[to] = findAvailableHop(next);
+          from = to;
+        }
+        if (to >= size) size = to + 1;
+      } else {
+        int rConflict = conflict(from, c);
+        int locate = base[rConflict] + c;
+        if (check[locate] != EPTY_BACK_VALUE) {
+          System.err.println("conflict");
+        }
+        check[locate] = rConflict;
+        if (i == klen - 1) {
+          base[locate] = -id;
+          id = id + 1;
+        } else {
+          int nah = bKeys[i + 1] & 0xFF;
+          base[locate] = findAvailableHop(nah);
+        }
+        if (locate >= size) size = locate + 1;
+        from = locate;
+        if (i == klen - 1) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Insert element (String) into DAT, the element will be transformed to
+   * byte[] firstly then insert into DAT.
+   *
+   * @param key
+   * @return
+   */
+  public boolean insert(String key) {
+    String k = key + '\0';
+    byte[] bKeys = k.getBytes();
+    if (!insert(bKeys)) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Serialize the DAT to data output stream
+   *
+   * @param out
+   * @throws IOException
+   */
+  public void write(DataOutputStream out) throws IOException {
```

Comment:
@kumarvishal09, thanks for your suggestions, I add serialization function since discussion about serializing dictionary at community, this may be convenient to further extend. FYI.